extensions-core/datasketches/src/main/java/io/druid/query/aggregation/datasketches/quantiles/DoublesSketchAggregatorFactory.java

```diff
@@ -0,0 +1,267 @@
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.druid.query.aggregation.datasketches.quantiles;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.metamx.common.IAE;
+import com.yahoo.sketches.Util;
+import com.yahoo.sketches.quantiles.DoublesSketch;
+import com.yahoo.sketches.quantiles.DoublesUnion;
+
+import io.druid.query.aggregation.Aggregator;
+import io.druid.query.aggregation.AggregatorFactory;
+import io.druid.query.aggregation.AggregatorFactoryNotMergeableException;
+import io.druid.query.aggregation.AggregatorUtil;
+import io.druid.query.aggregation.BufferAggregator;
+import io.druid.query.cache.CacheKeyBuilder;
+import io.druid.segment.ColumnSelectorFactory;
+import io.druid.segment.ColumnValueSelector;
+import io.druid.segment.column.ValueType;
+
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.List;
+
+public class DoublesSketchAggregatorFactory extends AggregatorFactory
+{
+
+  private static final int DEFAULT_K = 128;
+
+  // Used for sketch size estimation.
+  private static final long MAX_STREAM_LENGTH = 1_000_000_000;
+
+  private final String name;
+  private final String fieldName;
+  private final int k;
+  private final byte cacheTypeId;
+
+  @JsonCreator
+  public DoublesSketchAggregatorFactory(
+      @JsonProperty("name") final String name,
+      @JsonProperty("fieldName") final String fieldName,
+      @JsonProperty("k") final Integer k)
+  {
+    this(name, fieldName, k, AggregatorUtil.QUANTILES_DOUBLES_SKETCH_BUILD_CACHE_TYPE_ID);
+  }
+
+  DoublesSketchAggregatorFactory(final String name, final String fieldName, final Integer k, final byte cacheTypeId)
+  {
+    if (name == null) {
+      throw new IAE("Must have a valid, non-null aggregator name");
+    }
+    this.name = name;
+    if (fieldName == null) {
+      throw new IAE("Parameter fieldName must be specified");
+    }
+    this.fieldName = fieldName;
+    this.k = k == null ? DEFAULT_K : k;
+    Util.checkIfPowerOf2(this.k, "size");
+    this.cacheTypeId = cacheTypeId;
+  }
+
+  @Override
+  public Aggregator factorize(final ColumnSelectorFactory metricFactory)
+  {
+    if (metricFactory.getColumnCapabilities(fieldName) != null
+        && ValueType.isNumeric(metricFactory.getColumnCapabilities(fieldName).getType())) {
+      final ColumnValueSelector<Double> valueSelector = metricFactory.makeColumnValueSelector(fieldName);
+      if (valueSelector == null) {
+        return new DoublesSketchNoOpAggregator();
+      }
+      return new DoublesSketchBuildAggregator(valueSelector, k);
+    }
+    final ColumnValueSelector<DoublesSketch> selector = metricFactory.makeColumnValueSelector(fieldName);
+    if (selector == null) {
+      return new DoublesSketchNoOpAggregator();
+    }
+    return new DoublesSketchMergeAggregator(selector, k);
```

Comment:
This is a common entry point associated with the type name. If the input field is a numeric field, then so-called "build" aggregator is used to build sketches. Otherwise, we assume that the input contains sketches to merge.

Reply:
Druid calls ```combine()``` or gets combiningFactories by calling ```getCombiningFactory()```  for merging aggregates. Calling ```DoublesSketchAggregatorFactory.factorize()``` should not happen for merging aggregates.

Reply:
Perhaps, I  did not make myself clear. By merging sketches I meant such an aggregation in which the input field contains sketches as opposed to the raw values. I used to think about this as having two modes: building sketches (from raw input) and merging sketches. The closest thing to this is Theta sketch aggregator, with a twist that it cannot autodetect the input type (here the input is numeric, but there it can be of almost any type).

Reply:
Ah, sorry I misunderstood. Sounds good.